shutdown.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+)
+
+// Shutdowner provides a method that can manually trigger the shutdown of the
+// application by sending a signal to all open Done channels. Shutdowner works
+// both on applications using Run as well as Start, Done, and Stop.
+type Shutdowner interface {
+	Shutdown(...ShutdownOption) error
+}
+
+// Shutdown broadcasts a signal to all of the Done channels on the application
+// thereby enabling the start of the Stop process.
+func (s *shutdowner) Shutdown(opts ...ShutdownOption) error {
+	return s.broadcastSignal(syscall.SIGTERM)
+}
+
+// ShutdownOption provides a way to configure properties of the shutdown
+// process. Currently, no options have been implemented.
+type ShutdownOption interface {
+	apply(*shutdowner)
+}
+
+type shutdowner struct {
+	broadcastSignal func(os.Signal) error
+}
+
+func (app *App) shutdowner() Shutdowner {
+	return &shutdowner{
+		broadcastSignal: app.signalBroadcaster(),
+	}
+}
+
+func (app *App) signalBroadcaster() func(os.Signal) error {
+	return func(signal os.Signal) error {
+		app.mu.RLock()
+		defer app.mu.RUnlock()
+
+		var unsent int
+		for i, done := range app.dones {
+			select {
+			case done <- signal:
+			default:
+				// signal was never sent to channel
+				unsent++
+				app.logger.Printf("done channel %d at capacity, did not receive shutdown signal", i)
```

Comment:
I don't know if we need to record this at all. The contract for signal
handlers is if you're listening, you get the message, otherwise it's lost. The
same should apply here if we're emulating that.

Reply:
@amckinney lmk what you think of that comment
@abhinav I'm not committed to returning this as an error, necessarily, but I do think that it is valuable to indicate that the done channel already received a SIGTERM or SIGINT signal that has not yet been cleared from the channel (even if we can't provide a whole lot of context around what sent the blocking signal). I also agree with @akshayjshah that it makes sense to *either* log the error or return the error, and I also agree that it makes more sense to return the error than log it.

Reply:
@akshayjshah if we only return the error outside of the loop and forgo logging each blocked channel as we come across it, do we think it's still worthwhile to retain the index of each blocked channel and combine them into the single error message?

Reply:
Probably not. An error message like "failed to send SIGTERM to 4 of 7 channels" is good enough.